src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,991 @@
+  ldout(m_cct, 10) << dendl;
+
+  uint64_t size;
+  {
+    RWLock::RLocker snap_locker(m_image_ctx->snap_lock);
+    size = m_image_ctx->size;
+  }
+
+  ThreadPool *thread_pool;
+  ContextWQ *op_work_queue;
+  ImageCtx::get_thread_pool_instance(m_cct, &thread_pool, &op_work_queue);
+
+  C_SaferCond on_create;
+  auto *req = image::CreateRequest<I>::create(
+      m_dst_io_ctx, m_dst_image_name, m_dst_image_id, size, m_image_options, "",
+      "", false, op_work_queue, &on_create);
+  req->send();
+  int r = on_create.wait();
+  if (r < 0) {
+    lderr(m_cct) << "header creation failed: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  auto dst_image_ctx = new librbd::ImageCtx(m_dst_image_name, m_dst_image_id,
+                                            nullptr, m_dst_io_ctx, false);
+  dst_image_ctx->ignore_migrating = true; // XXXMG: hackish
+  r = dst_image_ctx->state->open(false);
+  if (r < 0) {
+    lderr(m_cct) << "failed to open newly created header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  dst_image_ctx->owner_lock.get_read();
+  if (dst_image_ctx->exclusive_lock != nullptr &&
+      !dst_image_ctx->exclusive_lock->is_lock_owner()) {
+    C_SaferCond ctx;
+    dst_image_ctx->exclusive_lock->acquire_lock(&ctx);
+    dst_image_ctx->owner_lock.put_read();
+    r = ctx.wait();
+    if (r < 0) {
+      lderr(m_cct) << "error acquiring exclusive lock: " << cpp_strerror(r)
+                   << dendl;
+      return r;
+    }
+  } else {
+    dst_image_ctx->owner_lock.put_read();
+  }
+
+  SnapSeqs snap_seqs;
+
+  C_SaferCond on_snapshot_copy;
+  auto snapshot_copy_req = librbd::deep_copy::SnapshotCopyRequest<I>::create(
+      m_image_ctx, dst_image_ctx, m_image_ctx->op_work_queue, &snap_seqs,
+      &on_snapshot_copy);
+  snapshot_copy_req->send();
+  r = on_snapshot_copy.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to copy snapshots: " << cpp_strerror(r) << dendl;
+    dst_image_ctx->state->close();
+    return r;
+  }
+
+  C_SaferCond on_set_head;
+  librbd::ParentSpec parent_spec = {};
+  uint64_t parent_overlap = 0;
+  auto set_head_req = librbd::deep_copy::SetHeadRequest<I>::create(
+      dst_image_ctx, size, parent_spec, parent_overlap, &on_set_head);
+  set_head_req->send();
+  r = on_set_head.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to set image head: " << cpp_strerror(r) << dendl;
+    dst_image_ctx->state->close();
+    return r;
+  }
+
+  r = dst_image_ctx->state->close();
+  if (r < 0) {
+    lderr(m_cct) << "failed to close image: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  cls::rbd::MigrationSpec migration_spec = {cls::rbd::MIGRATION_TYPE_DST,
+                                            cls::rbd::MIGRATION_STATE_STARTED,
+                                            m_src_io_ctx.get_id(),
+                                            m_src_image_name, m_src_image_id,
+                                            snap_seqs, size, false};
```

Comment:
Is it necessary to persist the snap sequence mapping? Right now, this is a linear process where if it is aborted mid-migration, you cannot restart it where it left off.

Reply:
I also forgot to answer this. I need to persist the snap sequence mapping so it could be available for "deep copyup" (see `CopyupRequest::send`).